File: leetcode_py/Path Sum II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def ins(self, root, target, path, res):
        if root is None:
            return res
        if root.val == target and root.left is None and root.right is None:
            return res + [path + [root.val]]
        else:
            return self.ins(root.left, target - root.val, path + [root.val], res) + \
                   self.ins(root.right, target - root.val, path + [root.val], res)

    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        return self.ins(root, sum, [], [])
        # return res


if __name__ == "__main__":
    a = Solution()
    a.ins()


class Solution:
    def ins(self, root, target, lists):
        if root is None:
            return
        if root.val == target and root.left is None and root.right is None:
            self.res.append(lists + [root.val])
        else:
            if root.left is not None:
                tmp = [i for i in lists]
                self.ins(root.left, target - root.val, tmp + [root.val])
            if root.right is not None:
                tmp = [i for i in lists]
                self.ins(root.right, target - root.val, tmp + [root.val])

    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        self.res = []
        if root is None:
            return []
        self.ins(root, sum, [])
        return self.res

File: leetcode_js/Path Sum II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
function helper(root, sum, path, res){
    path = path.concat(root.val);
    if(!root.left && !root.right){
        if(root.val === sum){
            res.push(path);
        }
        return;
    }
    if(root.left){
        helper(root.left, sum - root.val, path, res);
    }
     if(root.right){
        helper(root.right, sum - root.val, path, res);
    }
}
var pathSum = function(root, sum) {
    if(!root) return [];
    var res = [];
    helper(root, sum, [], res);
    return res;
};